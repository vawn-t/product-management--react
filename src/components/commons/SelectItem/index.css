.select-wrapper {
  padding: 1rem 1rem;
}

.select {
  border-radius: 3px;
  font-size: 16px;
  width: 100%;
  outline: none;
  padding: 0.5rem;
  text-indent: 0.25rem;
  background: var(--light-gray);
  border: 1px solid var(--light-gray);
  margin-top: 0.25rem;
}

.select:disabled {
  color: black;
  cursor: text;
  -webkit-touch-callout: text; /* iOS Safari */
  -webkit-user-select: text; /* Safari */
  -khtml-user-select: text; /* Konqueror HTML */
  -moz-user-select: text; /* Old versions of Firefox */
  -ms-user-select: text; /* Internet Explorer/Edge */
  user-select: text;
  opacity: 1;
}
