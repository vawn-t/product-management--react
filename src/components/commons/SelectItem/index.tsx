// Libraries
import React, {
  ChangeEvent,
  forwardRef,
  memo,
  useState,
  useCallback,
  useEffect
} from 'react';

// Styles
import './index.css';

interface SelectItemProps {
  name: string;
  defaultValue: string;
  label: string;
  options: string[];
  disable?: boolean;
  onSelectChange?: (value: string, fieldName?: string) => void;
}

const SelectItem = forwardRef<HTMLSelectElement, SelectItemProps>(
  (
    { name, defaultValue, label, options, disable = false, onSelectChange },
    ref = null
  ) => {
    const [value, setValue] = useState<string>(defaultValue);
    const [isFirstRun, setIsFirstRun] = useState(true);

    const handleChange = useCallback((event: ChangeEvent) => {
      const newValue = (event.currentTarget as HTMLInputElement).value;
      setValue(newValue);

      onSelectChange && onSelectChange(newValue);
    }, []);

    useEffect(() => {
      if (isFirstRun) {
        setIsFirstRun(false);
        return;
      }

      setValue(defaultValue);
    }, [defaultValue]);

    return (
      <div className='select-wrapper'>
        <label htmlFor={name}>{label}:</label>
        <select
          className='select'
          name={name}
          id={name}
          value={value}
          disabled={disable}
          onChange={handleChange}
          data-testid='select-option'
          ref={ref}
        >
          <option value=''>Select</option>
          {options.map((option: string, index: number) => (
            <option key={index} value={option}>
              {option}
            </option>
          ))}
        </select>
      </div>
    );
  }
);

SelectItem.displayName = 'SelectItem';

export default memo(SelectItem);
