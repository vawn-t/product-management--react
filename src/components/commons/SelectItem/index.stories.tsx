// Libraries
import React from 'react';
import { ComponentStory, ComponentMeta } from '@storybook/react';

// Components
import SelectItem from './index';

// Styles
import '@assets/styles/reset.css';
import '@assets/styles/App.css';
import '@assets/styles/variables.css';

// Constants
import { PRODUCT_TYPE_LIST } from '@constants/index';

export default {
  title: 'Components/SelectItem',
  component: SelectItem
} as ComponentMeta<typeof SelectItem>;

const Template: ComponentStory<typeof SelectItem> = (args) => (
  <SelectItem {...args} />
);

export const Default = Template.bind({});
Default.args = {
  label: 'SelectItem label',
  name: 'select',
  options: PRODUCT_TYPE_LIST
};
