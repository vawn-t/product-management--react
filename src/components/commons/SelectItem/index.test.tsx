// Libraries
import React from 'react';
import { render, fireEvent, cleanup, screen } from '@testing-library/react';
import { create } from 'react-test-renderer';
import '@testing-library/jest-dom';

// Components
import SelectItem from './index';

describe('Select Item component', () => {
  let container: HTMLElement;

  const defaultProps = {
    label: 'Select Item Label',
    name: 'select',
    options: ['option 1', 'option 2', 'option 3'],
    defaultValue: 'option 1',
    onSelectChange: jest.fn()
  };

  beforeEach(() => {
    // setup a DOM element as a render target
    container = document.createElement('div');
    document.body.appendChild(container);
  });

  afterEach(() => {
    // cleanup on exiting
    container.remove();
    jest.clearAllMocks();
    cleanup();
  });

  test('should render correctly', () => {
    const defaultSelectItem = create(<SelectItem {...defaultProps} />).toJSON();
    expect(defaultSelectItem).toMatchSnapshot();

    const selectItemWithValue = create(
      <SelectItem {...defaultProps} defaultValue={defaultProps.options[1]} />
    ).toJSON();
    expect(selectItemWithValue).toMatchSnapshot();
  });

  test('should render correct label', () => {
    const label = 'Select Component';
    render(<SelectItem {...defaultProps} label={label} />, { container });

    expect(container.querySelector('label')).toHaveTextContent(label);
  });
});
