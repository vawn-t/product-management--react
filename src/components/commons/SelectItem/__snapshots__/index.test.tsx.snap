// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Select Item component should render correctly 1`] = `
<div
  className="select-wrapper"
>
  <label
    htmlFor="select"
  >
    Select Item Label
    :
  </label>
  <select
    className="select"
    data-testid="select-option"
    disabled={false}
    id="select"
    name="select"
    onChange={[Function]}
    value="option 1"
  >
    <option
      value=""
    >
      Select
    </option>
    <option
      value="option 1"
    >
      option 1
    </option>
    <option
      value="option 2"
    >
      option 2
    </option>
    <option
      value="option 3"
    >
      option 3
    </option>
  </select>
</div>
`;

exports[`Select Item component should render correctly 2`] = `
<div
  className="select-wrapper"
>
  <label
    htmlFor="select"
  >
    Select Item Label
    :
  </label>
  <select
    className="select"
    data-testid="select-option"
    disabled={false}
    id="select"
    name="select"
    onChange={[Function]}
    value="option 2"
  >
    <option
      value=""
    >
      Select
    </option>
    <option
      value="option 1"
    >
      option 1
    </option>
    <option
      value="option 2"
    >
      option 2
    </option>
    <option
      value="option 3"
    >
      option 3
    </option>
  </select>
</div>
`;
